### PR TITLE
US-10 Floor Toggle

### PR DESCRIPTION
**Pull request by danseremet**
# Floor Toggle & 2 Indoor Maps

- Added a floor toggle for H building 8 & 9 floors.
- It supports future extension to other floors & buildings.
- Added 2 floor plans for toggling between (floor 8 and 9).
- Plus/minus button removed from the bottom right.
- Added special case for hide/show at a certain zoom threshold.
- Tests implemented for everything of this commit except what must use the IndoorMap class methods because it is extending OverlayView which is yet something to be determined on how to test because using methods from it results in not found methods by jasmine.

*Note:* Some screenshots are without the buildings overlay feature.
Peek 2020-02-29 00-57
Peek 2020-02-28 21-42
Peek 2020-02-28 20-26

### CODE REVIEW

File: src/app/core/components/indoor-map/indoor-map.component.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { AfterViewInit, Component, ElementRef, Input, ViewChild } from '@angular/core';
```
Comment (ClaudiaLapalme) [inline src/app/core/components/indoor-map/indoor-map.component.ts]: shouldn't this be on multiple lines?

File: src/app/core/components/indoor-map/indoor-map.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (ClaudiaLapalme) [inline src/app/core/components/indoor-map/indoor-map.component.spec.ts]: SHouldn't this be on multiple lines?

File: src/app/core/components/indoor-map/indoor-map.component.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+@Component({
+    selector: 'app-indoor-map',
+    templateUrl: './indoor-map.component.html',
+    styleUrls: ['./indoor-map.component.scss'],
+})
+export class IndoorMapComponent implements AfterViewInit {
+
+    @Input() map: google.maps.Map; // map reference
+
```
Comment (ClaudiaLapalme) [inline src/app/core/components/indoor-map/indoor-map.component.ts]: remove space

File: src/app/core/components/toggle-floors/toggle-floors.component.scss [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  font-size: 14px;
+  padding: 6px 9px 6px 9px;
+}
+
+.selected {
+  background-color: var(--ion-color-primary-contrast-rgb);
+  color: var(--ion-color-primary-shade);
+  font-size: 14px;
+  padding: 6px 9px 6px 9px;
+
```
Comment (ClaudiaLapalme) [inline src/app/core/components/toggle-floors/toggle-floors.component.scss]: remove space

File: src/app/core/components/toggle-floors/toggle-floors.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (ClaudiaLapalme) [inline src/app/core/components/toggle-floors/toggle-floors.component.spec.ts]: multiline

File: src/app/core/components/toggle-floors/toggle-floors.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import { IonicModule } from '@ionic/angular';
+
```
Comment (ClaudiaLapalme) [inline src/app/core/components/toggle-floors/toggle-floors.component.spec.ts]: remove space

File: src/app/core/components/toggle-floors/toggle-floors.component.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import { Component, OnInit, Output, EventEmitter, Input } from '@angular/core';
```
Comment (ClaudiaLapalme) [inline src/app/core/components/toggle-floors/toggle-floors.component.ts]: multiline

File: src/app/core/models/indoor-map.spec.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { IndoorMap } from "./indoor-map";
+import { ElementRef } from '@angular/core';
+
+describe('IndoorMap', () => {
+
+
```
Comment (ClaudiaLapalme) [inline src/app/core/models/indoor-map.spec.ts]: remove space

File: src/app/core/models/indoor-map.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    constructor(
+        public bounds: google.maps.LatLngBounds,
+        public map: google.maps.Map,
+        public div: any) {
+
+        super();
+
+        this.mapRef = map; // Necessary for hide/show behavior
+        this.divRef = div; // Necessary for hide/show behavior
+
```
Comment (ClaudiaLapalme) [inline src/app/core/models/indoor-map.ts]: remove space

File: src/app/core/models/indoor-map.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
```
Comment (ClaudiaLapalme) [inline src/app/core/models/indoor-map.ts]: remove space

File: src/app/core/models/indoor-map.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        public bounds: google.maps.LatLngBounds,
+        public map: google.maps.Map,
+        public div: any) {
+
+        super();
+
+        this.mapRef = map; // Necessary for hide/show behavior
+        this.divRef = div; // Necessary for hide/show behavior
+
+
```
Comment (ClaudiaLapalme) [inline src/app/core/models/indoor-map.ts]: this file has a lot of comments. Do we need all of them?

File: src/app/home/home.page.spec.ts [hunk 12]
```diff
@@ -17,9 +19,17 @@ describe('HomePage', () => {
         }
 
         TestBed.configureTestingModule({
-            declarations: [HomePage, ToggleCampusComponent],
+            declarations: [
+                HomePage,
+                ToggleCampusComponent,
```
Comment (ClaudiaLapalme) [inline src/app/home/home.page.spec.ts]: ToggleCampusComponent is written twice

File: src/app/home/home.page.ts [hunk 13]
```diff
@@ -17,44 +17,70 @@ export class HomePage implements AfterViewInit {
     @ViewChild('map', { static: false })
     mapElement: ElementRef;
 
+    // Reference to the native toggle campus html element
     @ViewChild('toggle', { read: ElementRef, static: false })
     toggle: ElementRef;
 
+    // Reference to the native switch floors buttons html element
```
Comment (ClaudiaLapalme) [inline src/app/home/home.page.ts]: is this comment necessary? it's repeating line 20

File: src/app/home/home.page.ts [hunk 14]
```diff
@@ -17,44 +17,70 @@ export class HomePage implements AfterViewInit {
...
     constructor(
         private mapService: MapService,
-    ) { 
+    ) {
         this.currentCenter = this.SGW;
+
+        // TODO: Remove initial indoorMap when we will be able to click
+        // on the building or zoom in close enough to switch
+        // from showing the building overlay to showing indoor maps.
+        this.indoorMapBuildingCode = 'H';    // initial buildingCode
```
Comment (ClaudiaLapalme) [inline src/app/home/home.page.ts]: the variable names should be self-explanatory. No need for comments

File: src/app/home/home.page.ts [hunk 15]
```diff
@@ -17,44 +17,70 @@ export class HomePage implements AfterViewInit {
...
+    }
 
     private loadMap(): void {
         this.mapService.loadMap(this.mapElement)
             .then(mapObj => {
-                this.map = mapObj;
+                this.mapModel = mapObj;     // Set the map reference to this component
+                this.mapLoaded = true;      // Necessary to pass mapLoaded to indoor maps after its loaded
+
+                // Toggle buttons
```
Comment (ClaudiaLapalme) [inline src/app/home/home.page.ts]: these comments are unnecessary

File: src/app/core/components/indoor-map/indoor-map.component.ts (hunk 1 above)
Comment (danseremet) [reply src/app/core/components/indoor-map/indoor-map.component.ts]: fixed

File: src/app/core/components/indoor-map/indoor-map.component.spec.ts (hunk 2 above)
Comment (danseremet) [reply src/app/core/components/indoor-map/indoor-map.component.spec.ts]: fixed

File: src/app/core/components/indoor-map/indoor-map.component.ts (hunk 3 above)
Comment (danseremet) [reply src/app/core/components/indoor-map/indoor-map.component.ts]: fixed

File: src/app/core/components/toggle-floors/toggle-floors.component.scss (hunk 4 above)
Comment (danseremet) [reply src/app/core/components/toggle-floors/toggle-floors.component.scss]: fixed

File: src/app/core/components/toggle-floors/toggle-floors.component.spec.ts (hunk 5 above)
Comment (danseremet) [reply src/app/core/components/toggle-floors/toggle-floors.component.spec.ts]: fixed

File: src/app/core/models/indoor-map.ts (hunk 11 above)
Comment (danseremet) [reply src/app/core/models/indoor-map.ts]: yes we do its important nobody touches them

File: src/app/home/home.page.ts (hunk 13 above)
Comment (danseremet) [reply src/app/home/home.page.ts]: not the same thing

File: src/app/home/home.page.ts (hunk 14 above)
Comment (danseremet) [reply src/app/home/home.page.ts]: fixed

File: src/app/home/home.page.ts (hunk 15 above)
Comment (danseremet) [reply src/app/home/home.page.ts]: fixed

File: src/app/home/home.page.spec.ts (hunk 12 above)
Comment (danseremet) [reply src/app/home/home.page.spec.ts]: fixed

File: src/app/core/components/indoor-map/indoor-map.component.ts (hunk 3 above)
Comment (ClaudiaLapalme) [reply src/app/core/components/indoor-map/indoor-map.component.ts]: why is the empty line still there?

**Comment by ClaudiaLapalme** [comment]
> Looks good to me.
> Will we be hiding the "toggle floors" button while we aren't looking at the inside of Hall?

For now that is a small bug and the work necessary to figure out how to fix it is not worth the time spent. Dan already spent a few hours trying to fix it and couldn't figure out how. We still retrieve the business value with this small bug in place, so I told him to forget about fixing this issue and to produce a PR. I will just log a bug report for it.

**Pull request merged**